core/src/server/access_control/cors.rs

```diff
@@ -239,38 +239,38 @@ pub(crate) fn get_cors_allow_origin(
 			}
 
 			match allowed.as_ref() {
-				None if *origin == "null" => AllowCors::Ok(AccessControlAllowOrigin::Null),
-				None => AllowCors::Ok(AccessControlAllowOrigin::Value(Origin::parse(origin))),
+				None if *origin == "null" => AllowCors::Ok(AllowOrigin::Null),
+				None => AllowCors::Ok(AllowOrigin::Origin(Origin::parse(origin))),
 				Some(allowed) if *origin == "null" => allowed
 					.iter()
-					.find(|cors| **cors == AccessControlAllowOrigin::Null)
+					.find(|cors| **cors == AllowOrigin::Null)
 					.cloned()
 					.map(AllowCors::Ok)
 					.unwrap_or(AllowCors::Invalid),
 				Some(allowed) => allowed
 					.iter()
 					.find(|cors| match **cors {
-						AccessControlAllowOrigin::Any => true,
-						AccessControlAllowOrigin::Value(ref val) if val.matches(origin) => true,
+						AllowOrigin::Any => true,
+						AllowOrigin::Origin(ref val) if val.matches(origin) => true,
 						_ => false,
 					})
-					.map(|_| AccessControlAllowOrigin::Value(Origin::parse(origin)))
+					.map(|_| AllowOrigin::Origin(Origin::parse(origin)))
 					.map(AllowCors::Ok)
 					.unwrap_or(AllowCors::Invalid),
 			}
 		}
 	}
 }
 
-/// Validates if the `AccessControlAllowedHeaders` in the request are allowed.
+/// Validates if the `AllowHeaders` in the request are allowed.
 pub(crate) fn get_cors_allow_headers<T: AsRef<str>, O, F: Fn(T) -> O>(
 	mut headers: impl Iterator<Item = T>,
 	requested_headers: impl Iterator<Item = T>,
-	cors_allow_headers: &AccessControlAllowHeaders,
+	cors_allow_headers: &AllowHeaders,
 	to_result: F,
 ) -> AllowCors<Vec<O>> {
 	// Check if the header fields which were sent in the request are allowed
-	if let AccessControlAllowHeaders::Only(only) = cors_allow_headers {
+	if let AllowHeaders::Only(only) = cors_allow_headers {
 		let are_all_allowed = headers.all(|header| {
 			let name = &Ascii::new(header.as_ref());
 			only.iter().any(|h| Ascii::new(&*h) == name) || ALWAYS_ALLOWED_HEADERS.contains(name)
```

Comment:
I'm struggling to understand `ALWAYS_ALLOWED_HEADERS`; I think that we allow any headers specified in AllowHeaders plus any in `ALWAYS_ALLOWED_HEADERS`.

I guess we have this list so that if the user doesn't allow any headers themselves via the access control stuff, standard requests will still work?

Some headers are also not completely black and white as to whether they are accepted or not. Eg with CORS, we'll always need to return "Content-Type" in the "Access-Control-Allowed-Headers" response, because otherwise I think the user couldn't set that to "application/json" without CORS disallowing it (see https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header#additional_restrictions).

Also there is an `Access-Control-Allow-Origin` header in the list, which is what the server would respond with, so perhaps this doesn't need to be there?
Also a client can send an `Access-Control-Request-Method` header in a preflight request, and so perhaps that should be there?

Reply:
Ok, so your suggestion is to only whitelist `Accept`, `Accept-Language`, `Content-Type` and `Content-Language`?